services/checks/checks.go

```diff
@@ -0,0 +1,170 @@
+	defaultInterval  = 24 * time.Hour
+
+	// Environment variables that affect checks service; only for testing.
+	envHost      = "PERCONA_TEST_CHECKS_HOST"
+	envPublicKey = "PERCONA_TEST_CHECKS_PUBLIC_KEY"
+	envInterval  = "PERCONA_TEST_CHECKS_INTERVAL"
+
+	timeout = 5 * time.Second
+)
+
+// Service is responsible for interactions with Percona Check service.
+type Service struct {
+	l          *logrus.Entry
+	pmmVersion string
+	host       string
+	publicKey  string
+	interval   time.Duration
+
+	m      sync.RWMutex
+	checks []check.Check
+}
+
+// New returns Service with given PMM version.
+func New(pmmVersion string) *Service {
+	l := logrus.WithField("component", "check")
+	s := &Service{
+		l:          l,
+		pmmVersion: pmmVersion,
+		host:       defaultHost,
+		publicKey:  defaultPublicKey,
+		interval:   defaultInterval,
+	}
+
+	if h := os.Getenv(envHost); h != "" {
+		l.Warnf("Host changed to %s.", h)
+		s.host = h
+	}
+	if k := os.Getenv(envPublicKey); k != "" {
+		l.Warnf("Public key changed to %s.", k)
+		s.publicKey = k
+	}
+	if i := os.Getenv(envInterval); i != "" {
+		l.Warnf("Interval changed to %s.", i)
+		s.publicKey = i
+	}
+
+	return s
+}
+
+// Run runs checks service that grabs checks from Percona Checks service every interval until context is canceled.
+func (s *Service) Run(ctx context.Context) {
+	ticker := time.NewTicker(s.interval)
+	defer ticker.Stop()
+
+	for {
+		if err := s.downloadChecks(ctx); err != nil {
+			s.l.Errorf("Failed to download checks: %v", err)
+		}
+
+		select {
+		case <-ticker.C:
+			// continue with next loop iteration
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Checks returns available checks.
+func (s *Service) Checks() []check.Check {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	r := make([]check.Check, 0, len(s.checks))
+	return append(r, s.checks...)
+}
+
+func (s *Service) downloadChecks(ctx context.Context) error {
+	s.l.Infof("Download checks from: %s", s.host)
+
+	host, _, err := net.SplitHostPort(s.host)
+	if err != nil {
+		return errors.Wrap(err, "failed to set telemetry host")
+	}
+	tlsConfig := tlsconfig.Get()
+	tlsConfig.ServerName = host
+
+	opts := []grpc.DialOption{
+		// replacement is marked as experimental
+		grpc.WithBackoffMaxDelay(timeout), //nolint:staticcheck
+		grpc.WithBlock(),
+		grpc.WithUserAgent("pmm-managed/" + s.pmmVersion),
+		grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)),
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	cc, err := grpc.DialContext(ctx, s.host, opts...)
+	if err != nil {
+		return errors.Wrap(err, "failed to dial")
+	}
+	defer cc.Close() //nolint:errcheck
+
+	resp, err := api.NewCheckedAPIClient(cc).GetAllChecks(ctx, &api.GetAllChecksRequest{})
+	if err != nil {
+		return errors.Wrap(err, "failed to request checks service")
+	}
+
+	if len(resp.Signatures) != 1 {
+		return errors.Errorf("wrong checks signatures number, expect one signature, received %d", len(resp.Signatures))
```

Comment:
Nope, we can receive several signatures. We should check all signatures with all hard-coded public keys. At least one should be verified. That will allow use to rotate keys every PMM release.

Reply:
Removed, added check for non-empty signatures slice